### PR TITLE
reliably determine available VRAM on macOS (resolves #1826, #2370)

### LINKED ISSUE #1826: MacOS: Ollama ignores changes to the iogpu.wired_limit_mb tunable when deciding whether to run on GPU or CPU
MacOS 14.2.1 on a 32GB M1 Max MBP
```
% ollama run dolphin-mixtral:8x7b-v2.7-q3_K_M
Error: model requires at least 48 GB of memory
```

This error appears immediately, it does not seem to try to load the model.

I tried pulling the model again. Same behavior. I've been running this model without issue on 0.1.17.

I tried upping the memory MacOS makes available to the GPU but it didn't help
`sudo sysctl iogpu.wired_limit_mb=26624`

Also an issue with mixtral:8x7b-instruct-v0.1-q3_K_M. nous-hermes2:34b-yi-q3_K_M runs, as does nous-hermes2:34b.

On 0.1.18,  nous-hermes2:34b's memory requirements, according to final `ggml_metal_add_buffer:` entry in the log, is
19675.33 MB and 21845.34 MB are available to the GPU

On 0.1.17, dolphin-mixtral:8x7b-v2.7-q3_K_M's 
19964.30 MB

On 0.1.17 mixtral:8x7b-instruct-v0.1-q3_K_M:
19965.17 MB

So, 0.1.18 runs a model that seems to require more memory than the q3_K_M mixtral variants that it refuses to run.

Has the memory requirement for the mixtral models increased dramatically in 0.1.18, or is this new feature of estimating and enforcing memory requirements causing problems?

### PR DESCRIPTION
**Pull request by peanut256**

### CODE REVIEW

**Comment by peanut256** [comment]
I improved the bugfix to solve #2370 without explicitly setting iogpu.wired_limit_mb

File: gpu/gpu_info_darwin.m [hunk 1]
```diff
@@ -0,0 +1,11 @@
+//go:build darwin
+#include "gpu_info_darwin.h"
+
+uint64_t getRecommendedMaxVRAM()
+{
+	id<MTLDevice> device = MTLCreateSystemDefaultDevice();
+	NSLog(@"Metal default device %s", [device.name UTF8String]);
```
Comment (jmorganca) [inline gpu/gpu_info_darwin.m:7]: Possible to remove the logging here as it's macOS-specific?

**Comment by jmorganca** [comment]
This looks great, thanks for the PR! Small comment on logging.

File: gpu/gpu_info_darwin.m (hunk 1 above)
Comment (peanut256) [reply gpu/gpu_info_darwin.m]: Sure, the maximal VRAM is now logged via log/slog.

File: gpu/gpu_darwin.go [hunk 2]
```diff
@@ -15,19 +19,9 @@ func CheckVRAM() (int64, error) {
...
-	// TODO: handle case where iogpu.wired_limit_mb is set to a higher value
-	if systemMemory <= 36*1024*1024*1024 {
-		systemMemory = systemMemory * 2 / 3
-	} else {
-		systemMemory = systemMemory * 3 / 4
-	}
-
-	return systemMemory, nil
+	recommendedMaxVRAM := int64(C.getRecommendedMaxVRAM())
+	slog.Info(fmt.Sprintf("Recommended max VRAM: %d MB", recommendedMaxVRAM/1024/1024))
```
Comment (jmorganca) [inline gpu/gpu_darwin.go:23]: Thanks! I think we'd want to log it here https://github.com/ollama/ollama/blob/main/llm/llm.go#L48 and let the caller of this function "decide" whether it should be logged

Comment (peanut256) [reply gpu/gpu_darwin.go]: Okay, then I remove logging from gpu_darwin.go.

I also checked twice whether I got the memory management correctly. Turns out that Core Foundation objects still need manual releasing in Objective-C, so I fixed that.

File: gpu/gpu_info_darwin.m [hunk 3]
```diff
@@ -0,0 +1,11 @@
+//go:build darwin
+#include "gpu_info_darwin.h"
+
+uint64_t getRecommendedMaxVRAM()
+{
+	id<MTLDevice> device = MTLCreateSystemDefaultDevice();
```
Comment (jmorganca) [inline gpu/gpu_info_darwin.m:6]: This is a side note but is this function callable from C? if so we might be able to do this all in pure Cgo in `gpu/gpu_darwin.go`. I haven't had a chance to test that.

Comment (peanut256) [reply gpu/gpu_info_darwin.m:6]: Apple only provides bindings for Swift, Objective-C and C++. It looks like we need a thin layer of Objective-C.

Comment (jmorganca) [reply gpu/gpu_info_darwin.m:6]: Great! Thanks for checking!

**Comment by jmorganca** [comment]
Tested this and it works great. Thanks for tracking this down. @easp and others will be very happy 😊

**Review by jmorganca — APPROVED** [review]

**Pull request merged**